### DISCUSSION TITLE
Hooks - How can I get data from the relationships?

### DISCUSSION THREAD

**Discussion by abdonrd** [directus/directus#11422] [Archive]
I have a hook like this:

`extensions/hooks/generate-slugs/index.js`
```javascript
const slugify = require('@sindresorhus/slugify');

module.exports = ({ filter }) => {
  filter('competitions.items.create', async (input) => {
    // I want to get the parent data
    // input.parent it's an ID from relationship
    const parent = null;

    return {
      ...input,
      slug: slugify(`${parent.name}-${input.name}`)
    };
  });
};
```

I want to get the `parent` data, because `input.parent` it's an ID from relationship.

**Comment by abdonrd** [comment]
I see that I can do this:

```javascript
const slugify = require('@sindresorhus/slugify');

module.exports = ({ filter }, { services }) => {
  const { ItemsService } = services;

  filter('competitions.items.create', async (input, { collection }, { database, schema }) => {
    const collectionService = new ItemsService(collection, { knex: database, schema });
    const parent = await collectionService.readOne(input.parent);

    return {
      ...input,
      slug: slugify(`${parent.name}-${input.name}`)
    };
  });
};
```

Is it necessary to send the `{ knex: database, schema }` to the `ItemsService`?
Without knowing much Directus, I expected something like:
```javascript
const collectionService = new ItemsService(collection);
```

**Comment by abdonrd** [comment]
Where can we find documentation about `ItemsService`?

**Reply by azrikahar** [reply]
It's definitely far from ideal, but the current best bet would be to look at it (and other services) from the source here: https://github.com/directus/directus/tree/main/api/src/services 👍

As for an actual documentation for it, we are tracking it in this issue: https://github.com/directus/directus/issues/10523

**Reply by abdonrd** [reply]
Ok, thanks for the references!

**Comment by rijkvanzanten** [comment]
> Is it necessary to send the { knex: database, schema } to the ItemsService?

Yes. The API internals reuse the existing connections and extracted schema information, otherwise, every service would have to redo the schema introspection, which would severely slow down the api 👍🏻

**Reply by abdonrd** [reply]
Ok, thanks @rijkvanzanten!
